### PR TITLE
Add table formatting on paste of detected table

### LINKED ISSUE #941: Paste table text as markdown table
**Is your feature request related to a problem? Please describe.**

In GoogleSheets/LibreOfficeCalc/MSExcel you can copy table (sections). You can't simply drop them into hedgedoc though.

**Describe the solution you'd like**

When pasting text into the editor, hedgedoc checks if it is formatted like a table and offers to re-write it to look like a real markdown table instead of text with TABs.

**Describe alternatives you've considered**

- Button/Menu-Item "Paste clipboard as table"
- Context-Menu-Item "Try to format as table" when selecting text

**Additional context**

### PR DESCRIPTION
**Pull request by ErikMichelson**
### Component/Part
Markdown editor

### Description
This PR adds support for pasting tables (e.g. from LibreOffice Calc or MS Excel) into HedgeDoc while reformatting them to markdown tables.

### Steps

- [x] Added implementation
- [x] Added / updated tests
- [x] Extended changelog
- [x] I read the [contribution documentation](https://github.com/hedgedoc/react-client/blob/main/CONTRIBUTING.md) and signed-off my commits to accept the DCO.

### Related Issue(s)
Closes #941

### CODE REVIEW

File: src/components/editor/editor-pane/editor-pane.tsx [hunk 1]
```diff
@@ -62,15 +63,31 @@ const onChange = (editor: Editor) => {
...
 
 const onPaste = (pasteEditor: Editor, event: PasteEvent) => {
-  if (event && event.clipboardData && event.clipboardData.files && event.clipboardData.files.length > 0) {
+  if (!event || !event.clipboardData) {
+    return
+  }
+
+  // Check for pasted tables
+  const pasteText = event.clipboardData.getData('text')
+  if (pasteText && pasteText.includes('\t') && pasteText.includes('\n')) {
```
Comment (mrdrogdrog) [inline src/components/editor/editor-pane/editor-pane.tsx:83]: This is a bit unsafe because these conditions could also apply to python code.
How about an "insert as table" action in the toolbar instead?

Comment (DerMolly) [reply src/components/editor/editor-pane/editor-pane.tsx:83]: We could also check if the code is pasted in between three backticks to prevent python code pasteing…

Also could we check that the code has the same number of tabs in each line?

Comment (mrdrogdrog) [reply src/components/editor/editor-pane/editor-pane.tsx:83]: sry but if we don't have a rock solid proof detection AND a possibility to say "insert as plain text" or "no, that's not what i meant! format as plain text instead (like in word. i love this feature)", i'm against a auto detection of stuff.

**Review by DerMolly — APPROVED** [review]

**Review by DerMolly — APPROVED** [review]

**Comment by DerMolly** [comment]
Maybe a toast would be good, to tell the user that you autoformated their stuff as a table and how to deactivate that?

**Comment by mrdrogdrog** [comment]
> Maybe a toast would be good, to tell the user that you autoformated their stuff as a table and how to deactivate that?

Did someone say "notifications"? 🌚

File: src/components/editor-page/editor-pane/editor-pane.tsx [hunk 2]
```diff
@@ -103,6 +98,29 @@ export const EditorPane: React.FC<EditorPaneProps & ScrollProps> = ({ onContentC
...
+    }
+
+    // Upload pasted files
+    if (event.clipboardData.files && event.clipboardData.files.length > 0) {
+      event.preventDefault()
+      const files: FileList = event.clipboardData.files
+      if (files && files.length >= 1) {
+        handleUpload(files[0], pasteEditor)
+      }
+    }
```
Comment (mrdrogdrog) [inline src/components/editor-page/editor-pane/editor-pane.tsx:121]: Instead of commenting I would extract these code blocks into separate functions

File: src/components/editor-page/editor-pane/table-extractor.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  // Tab-indented text should not be matched as a table
+  if (lines.every(line => line.startsWith('\t'))) {
+    return false
+  }
+  // Every line should have the same amount of tabs (table columns)
+  const tabsPerLines = lines.map(line => line.match(/\t/g)?.length ?? 0)
+  return tabsPerLines.every(line => line === tabsPerLines[0])
+}
+
+export const extractTable = (pasteData: string, pasteEditor: Editor): void => {
```
Comment (mrdrogdrog) [inline src/components/editor-page/editor-pane/table-extractor.ts:33]: This method is not only extracting the table, it also insert it into the editor. The name should reflect this. BUT, if you follow clean code, then you should know, that an "and" in the function-name is an indicator for "too much stuff is happening here". So you shouldn't insert the content into the editor here. This also increases the testability.

File: src/components/editor-page/editor-pane/editor-pane.tsx [hunk 4]
```diff
@@ -103,6 +98,29 @@ export const EditorPane: React.FC<EditorPaneProps & ScrollProps> = ({ onContentC
...
+      return
+    }
+
+    // Check for pasted tables
+    const pasteText = event.clipboardData.getData('text')
+    if (smartPasteEnabled && pasteText && isTable(pasteText)) {
+      event.preventDefault()
+      extractTable(pasteText, pasteEditor)
+      return
+    }
```
Comment (mrdrogdrog) [inline src/components/editor-page/editor-pane/editor-pane.tsx:112]: https://user-images.githubusercontent.com/6124140/107889213-4a809500-6f11-11eb-905f-c5f350f87e5c.mp4

File: src/components/editor-page/editor-pane/table-extractor.test.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
+/*
+ * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { isTable } from './table-extractor'
+
+describe('isTable detection', () => {
```
Comment (mrdrogdrog) [inline src/components/editor-page/editor-pane/table-extractor.test.ts:9]: Where are the tests for `extractTable` ?

File: cypress/integration/fileUpload.spec.ts [hunk 6]
```diff
@@ -45,40 +45,37 @@ describe('File upload', () => {
...
-          cy.get('.CodeMirror-scroll')
-            .trigger('drop', dropEvent)
-          cy.get('.CodeMirror-activeline > .CodeMirror-line > span')
-            .should('have.text', `![](${ imageUrl })`)
-        })
+        }
+        cy.get('.CodeMirror-scroll').trigger('dragenter', dropEvent)
+        cy.get('.CodeMirror-scroll').trigger('drop', dropEvent)
+        cy.get('.CodeMirror-activeline > .CodeMirror-line > span')
+          .should('have.text', `![](${imageUrl})`)
```
Comment (mrdrogdrog) [inline cypress/integration/fileUpload.spec.ts:77]: This seems to be malformatted, because these are chained function calls.

File: cypress/integration/fileUpload.spec.ts [hunk 7]
```diff
@@ -45,40 +45,37 @@ describe('File upload', () => {
       cy.get('div.btn-group > input[type=file]')
         .attachFile({ filePath: 'acme.png', mimeType: 'image/png' })
       cy.get('.CodeMirror-activeline > .CodeMirror-line > span')
-        .should('have.text', `![](${ imageUrl })`)
+        .should('have.text', `![](${imageUrl})`)
```
Comment (mrdrogdrog) [inline cypress/integration/fileUpload.spec.ts:48]: This seems to be malformatted, because according to our new editor config templates should have white spaces around the content.

**Review by mrdrogdrog — CHANGES_REQUESTED** [review]

**Comment by mrdrogdrog** [comment]
Sry but I'm still not fully convinced regarding the usability of this feature as long as I can't say "No! I don't want this text to be autoformated. Revert that!"

File: src/components/editor-page/editor-pane/editor-pane.tsx (hunk 4 above)
Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/editor-pane.tsx:112]: What's your idea how to detect whether the cursor is inside a code-fence or not?

File: cypress/integration/fileUpload.spec.ts (hunk 7 above)
Comment (ErikMichelson) [reply cypress/integration/fileUpload.spec.ts:48]: Hm, I thought I ran the "reformat code" method.

Fixed.

File: cypress/integration/fileUpload.spec.ts (hunk 6 above)
Comment (ErikMichelson) [reply cypress/integration/fileUpload.spec.ts:77]: Changed

**Comment by ErikMichelson** [comment]
You might have seen that there is the option to disable this kind of paste-reformatting. But you still want a "revert that" button. Am I right?

If so, we'd need to ensure to replace only the reformatted table (which could have been changed already by someone else) in the document with the plain text and we'd need to store the original paste content somewhere. Sounds not quite trivial.

File: src/components/editor-page/editor-pane/editor-pane.tsx (hunk 4 above)
Comment (mrdrogdrog) [reply src/components/editor-page/editor-pane/editor-pane.tsx:112]: Check the line above and below (except blank lines)?

**Comment by mrdrogdrog** [comment]
Tbh.. yes. This is what I would want, because it could confuse the user if they don't know how to disable the feature and if that what they insert is not what they expect. For now this is okay, but imho this would be an excellent candidate for the notifications.

**Comment by jfowl** [comment]
I know you did a lot of work on this already, but what about a button/shortcut which attempts to transform the selected text into a table. sou you would paste your table, select it and then press a button to get a table. Could also be used to create table from CSV code blocks. And it would be explicit.

**Comment by mrdrogdrog** [comment]
I'm not sure if this would be more or less confusing for users who copied a table and expect to insert a table.

**Comment by mrdrogdrog** [comment]
My problem is, that you can't easily revert the paste if the algorithm false-positive detected a table.

**Comment by jfowl** [comment]
What about a modal popping up when pasting to ask you?

**Comment by mrdrogdrog** [comment]
That would interrupt your workflow. Modals which pop up out of nowhere are bad ux.

**Comment by mrdrogdrog** [comment]
Like I said: For now this is .. okay. But we should test test test this. As you can see, unhandled cases (like the paste in the code block) can happen every time.

I still think this would be a great candidate for our notification/toast-idea

File: src/components/editor-page/editor-pane/editor-pane.tsx (hunk 2 above)
Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/editor-pane.tsx]: Done

File: src/components/editor-page/editor-pane/table-extractor.ts (hunk 3 above)
Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/table-extractor.ts]: Changed

File: src/components/editor-page/editor-pane/editor-pane.tsx (hunk 4 above)
Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/editor-pane.tsx]: I now did the following: Check if either the cursor's line, the one above or the one below have the `m-markdown comment` CodeMirro token, which is applied to codefence blocks.
I need to check the line above and below too as empty lines - although being inside the fence - have no CodeMirror token.

From this results the following:
 1. Reformatting won't trigger if the line above or below starts with `` ``` `` or is inside the fence and has any content.
 2. Reformatting will trigger if you're outside the codefence but directly above the fence-start (next line begins with `` ``` ``).
 3. Same as 2. when you're outside a codefence but the previous line is the end of it.

File: src/components/editor-page/editor-pane/table-extractor.test.ts (hunk 5 above)
Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/table-extractor.test.ts]: Added

File: src/components/editor-page/editor-pane/table-extractor.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+                             .filter(row => row.trim() !== '')
+  const tableCells = tableRows.reduce((cellsInRow, row, index) => {
+    cellsInRow[index] = row.split('\t')
+    return cellsInRow
+  }, [] as string[][])
+  const arrayMaxRows = createNumberRangeArray(tableCells.length)
+  const arrayMaxColumns = createNumberRangeArray(Math.max(...tableCells.map(row => row.length)))
+
+  const headRow1 = arrayMaxColumns
+    .map(col => col + 1)
```
Comment (mrdrogdrog) [inline src/components/editor-page/editor-pane/table-extractor.ts]: :banana: I think you can spare this step and use the incremented value in the next step

File: src/components/editor-page/editor-pane/table-extractor.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    .map(col => `| #${ col } `)
+    .join('') + '|'
+  const headRow2 = arrayMaxColumns
+    .map(col => col + 1)
```
Comment (mrdrogdrog) [inline src/components/editor-page/editor-pane/table-extractor.ts]: :banana: same here

File: src/components/editor-page/editor-pane/table-extractor.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    cellsInRow[index] = row.split('\t')
+    return cellsInRow
+  }, [] as string[][])
+  const arrayMaxRows = createNumberRangeArray(tableCells.length)
+  const arrayMaxColumns = createNumberRangeArray(Math.max(...tableCells.map(row => row.length)))
+
+  const headRow1 = arrayMaxColumns
+    .map(col => col + 1)
+    .map(col => `| #${ col } `)
+    .join('') + '|'
```
Comment (mrdrogdrog) [inline src/components/editor-page/editor-pane/table-extractor.ts:43]: Why don't you join the column texts with " | " ?

File: src/components/editor-page/editor-pane/table-extractor.test.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    const input = '1\t1\n\t0\n\t0\n4\t16\n5\t25\n'
+    expect(isTable(input))
+      .toBe(true)
+  })
+
+  it('valid table with some last cells missing', () => {
+    const input = '1\t1\n2\t\n3\t\n4\t16\n'
+    expect(isTable(input))
+      .toBe(true)
+  })
```
Comment (mrdrogdrog) [inline src/components/editor-page/editor-pane/table-extractor.test.ts:63]: These case names are different from the others. I'm missing the expectation.

File: src/components/editor-page/editor-pane/table-extractor.test.ts [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+})
+
+describe('Conversion from "excel" to markdown table', () => {
```
Comment (mrdrogdrog) [inline src/components/editor-page/editor-pane/table-extractor.test.ts:66]: How about "Tab and newline separated" instead of "excel"?

File: src/components/editor-page/editor-pane/table-extractor.test.ts [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+    const input = '1\t1\n\t0\n\t0\n4\t16\n'
+    expect(excelTableToMarkdown(input))
+      .toEqual('| #1 | #2 |\n| -- | -- |\n| 1 | 1 |\n|  | 0 |\n|  | 0 |\n| 4 | 16 |')
+  })
+
+  it('works with table with some last cells missing', () => {
+    const input = '1\t1\n2\t\n3\t\n4\t16\n'
+    expect(excelTableToMarkdown(input))
+      .toEqual('| #1 | #2 |\n| -- | -- |\n| 1 | 1 |\n| 2 |  |\n| 3 |  |\n| 4 | 16 |')
+  })
```
Comment (mrdrogdrog) [inline src/components/editor-page/editor-pane/table-extractor.test.ts:89]: You're only testing positive cases. Aren't there any edge cases?

**Review by mrdrogdrog — CHANGES_REQUESTED** [review]

File: src/components/editor-page/editor-pane/table-extractor.ts (hunk 8 above)
Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/table-extractor.ts]: Done

File: src/components/editor-page/editor-pane/table-extractor.ts (hunk 9 above)
Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/table-extractor.ts]: Done

File: src/components/editor-page/editor-pane/table-extractor.test.ts (hunk 12 above)
Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/table-extractor.test.ts:66]: I thought about names and they were all so long.... :D

File: src/components/editor-page/editor-pane/table-extractor.test.ts (hunk 11 above)
Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/table-extractor.test.ts:63]: What I'll do: Remove the _"valid"_ at the beginning and append _"is valid"_ at the end.

File: src/components/editor-page/editor-pane/table-extractor.test.ts (hunk 12 above)
Comment (mrdrogdrog) [reply src/components/editor-page/editor-pane/table-extractor.test.ts:66]: Yes but "Excel" is a bit to vague.

**Review by mrdrogdrog — CHANGES_REQUESTED** [review]

File: src/components/editor-page/editor-pane/table-extractor.ts (hunk 10 above)
Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/table-extractor.ts:43]: What would be the benefit? It only requires an additional `|` at the beginning.

Comment (mrdrogdrog) [reply src/components/editor-page/editor-pane/table-extractor.ts:43]: You have one string formating step (the template in the map) less?

File: src/components/editor-page/editor-pane/table-extractor.test.ts (hunk 13 above)
Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/table-extractor.test.ts:89]: To be honest: I can't think of one. Maybe I'm just too uncreative - do you see any edge case I should add?

File: src/components/editor-page/editor-pane/table-extractor.test.ts (hunk 12 above)
Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/table-extractor.test.ts]: Changed

File: src/components/editor-page/editor-pane/table-extractor.ts (hunk 10 above)
Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/table-extractor.ts]: Changed when applicable

Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/table-extractor.ts]: This change only makes things harder. I forgot the `#` in the template literal.
I'll revert my change back to the version you reviewed.

File: src/components/editor-page/editor-pane/table-extractor.ts [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+  // Tab-indented text should not be matched as a table
+  if (lines.every(line => line.startsWith('\t'))) {
+    return false
+  }
+  // Every line should have the same amount of tabs (table columns)
+  const tabsPerLines = lines.map(line => line.match(/\t/g)?.length ?? 0)
+  return tabsPerLines.every(line => line === tabsPerLines[0])
+}
+
+export const clipboardTableToMarkdown = (pasteData: string): string => {
```
Comment (mrdrogdrog) [inline src/components/editor-page/editor-pane/table-extractor.ts:31]: Clipboard is a weird verb :yum:

File: src/components/editor-page/editor-pane/table-extractor.test.ts [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+  it('works with normal table with newline at end', () => {
+    const input = '1\t1\n2\t4\n3\t9\n4\t16\n'
+    expect(clipboardTableToMarkdown(input))
+      .toEqual('| #1 | #2 |\n| -- | -- |\n| 1 | 1 |\n| 2 | 4 |\n| 3 | 9 |\n| 4 | 16 |')
+  })
+
+  it('works with table with some first cells missing', () => {
+    const input = '1\t1\n\t0\n\t0\n4\t16\n'
```
Comment (mrdrogdrog) [inline src/components/editor-page/editor-pane/table-extractor.test.ts:80]: These two blocks have different naming pattern. Can we please agree on a pattern? :sweat_smile:

File: src/components/editor-page/editor-pane/table-extractor.ts (hunk 10 above)
Comment (mrdrogdrog) [reply src/components/editor-page/editor-pane/table-extractor.ts:43]: :shrug:

File: src/components/editor-page/editor-pane/table-extractor.test.ts (hunk 13 above)
Comment (mrdrogdrog) [reply src/components/editor-page/editor-pane/table-extractor.test.ts]: What about empty string?

File: src/components/editor-page/editor-pane/table-extractor.ts (hunk 14 above)
Comment (DerMolly) [reply src/components/editor-page/editor-pane/table-extractor.ts:31]: May I suggest `convertCclipboardTableToMarkdown`

Comment (mrdrogdrog) [reply src/components/editor-page/editor-pane/table-extractor.ts:31]: nope. to many carbon copies.

Comment (mrdrogdrog) [reply src/components/editor-page/editor-pane/table-extractor.ts:31]: `cC`

Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/table-extractor.ts]: Changed

File: src/components/editor-page/editor-pane/table-extractor.test.ts (hunk 15 above)
Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/table-extractor.test.ts:80]: Changed

File: src/components/editor-page/editor-pane/table-extractor.test.ts (hunk 13 above)
Comment (ErikMichelson) [reply src/components/editor-page/editor-pane/table-extractor.test.ts]: An empty string won't ever be passed to this method. The method is called only when `isTable(...)` returns true. But that won't return true for an empty string.

Nevertheless, because I know you'd insist on this, I've added a relevant case. :D

**Review by mrdrogdrog — CHANGES_REQUESTED** [review]
You're still not ignoring empty lines in code blocks.

**Review by mrdrogdrog — APPROVED** [review]
